File: DevExpress-Demo/CS/DockingDemo.Wpf/Helpers/DashboardDataHelper.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Reflection;
using DevExpress.DemoData.Helpers;
using DevExpress.Xpf.DemoBase.Helpers;

namespace DockingDemo.Helpers {
    public static class DataLoader {
        static DataSet LoadData(string fileName) {
            string path = DataFilesHelper.FindFile(string.Format("{0}.xml", fileName), DataFilesHelper.DataPath);
            DataSet ds = new DataSet();
            ds.ReadXml(path, XmlReadMode.ReadSchema);
            return ds;
        }
        public static DataSet LoadSales() {
            return LoadData("DashboardSales");
        }
    }
    public abstract class SalesDataGenerator {
        public class Context {
            readonly string st;
            readonly string prodName;
            readonly string catName;
            readonly decimal lPrice;
            readonly UnitsSoldRandomGenerator uSoldGenerator;

            public string State { get { return st; } }
            public string ProductName { get { return prodName; } }
            public string CategoryName { get { return catName; } }
            public decimal ListPrice { get { return lPrice; } }
            public UnitsSoldRandomGenerator UnitsSoldGenerator { get { return uSoldGenerator; } }

            public Context(string st, string prodName, string catName, decimal lPrice, UnitsSoldRandomGenerator uSoldGenerator) {
                this.st = st;
                this.prodName = prodName;
                this.catName = catName;
                this.lPrice = lPrice;
                this.uSoldGenerator = uSoldGenerator;
            }
        }

        protected static string GetState(DataRow region) {
            return (string)region["Region"];
        }
        protected static string GetProductName(DataRow product) {
            return (string)product["Name"];
        }
        protected static decimal GetListPrice(DataRow product) {
            return (decimal)product["ListPrice"];
        }

        readonly DataSet ds;
        readonly DataTable categoriesTable;
        readonly DataTable productsTable;
        readonly DataTable regionsTable;
        readonly Random rand = new Random(1);
        readonly ProductClasses prodClasses;
        readonly RegionClasses regClasses;

        protected DataRowCollection Regions { get { return regionsTable.Rows; } }
        protected DataRowCollection Products { get { return productsTable.Rows; } }
        protected ProductClasses ProdClasses { get { return prodClasses; } }
        protected RegionClasses RegClasses { get { return regClasses; } }
        protected Random Random { get { return rand; } }

        protected SalesDataGenerator(DataSet ds) {
            this.ds = ds;
            categoriesTable = ds.Tables["Categories"];
            productsTable = ds.Tables["Products"];
            regionsTable = ds.Tables["Regions"];
            prodClasses = new ProductClasses(productsTable.Rows);
            regClasses = new RegionClasses(regionsTable.Rows);
        }
        protected double GetRegionWeigtht(DataRow region) {
            return regClasses[(int)region["RegionID"]];
        }
        protected ProductClass GetProductClass(DataRow product) {
            return prodClasses[(int)product["ProductID"]];
        }
        protected string GetCategoryName(DataRow product) {
            return (string)categoriesTable.Select(string.Format("CategoryID = {0}", product["CategoryID"]))[0]["CategoryName"];
        }
        protected UnitsSoldRandomGenerator CreateUnitsSoldGenerator(double regionWeight, ProductClass productClass) {
            return new UnitsSoldRandomGenerator(rand, (int)Math.Ceiling(productClass.SaleProbability * regionWeight));
        }
        protected abstract void Generate(Context context);
        protected virtual void EndGenerate() {
        }
        public void Generate() {
            foreach(DataRow region in Regions) {
                string state = GetState(region);
                double regionWeight = GetRegionWeigtht(region);
                foreach(DataRow product in Products) {
                    UnitsSoldRandomGenerator unitsSoldgenerator = CreateUnitsSoldGenerator(regionWeight, GetProductClass(product));
                    Generate(new Context(state, GetProductName(product), GetCategoryName(product), GetListPrice(product), unitsSoldgenerator));
                }
            }
            EndGenerate();
        }
    }
    public class SalesPerformanceDataGenerator : SalesDataGenerator {
        readonly List<MonthlySalesItem> monthlySales = new List<MonthlySalesItem>();
        readonly List<TotalSalesItem> totalSales = new List<TotalSalesItem>();
        readonly KeyMetricsItem item = new KeyMetricsItem();

        public IEnumerable<MonthlySalesItem> MonthlySales { get { return monthlySales; } }
        public IEnumerable<TotalSalesItem> TotalSales { get { return totalSales; } }
        public IEnumerable<KeyMetricsItem> KeyMetrics { get { return new KeyMetricsItem[] { item }; } }

        public SalesPerformanceDataGenerator(DataSet dataSet)
            : base(dataSet) {
        }
        protected override void Generate(Context context) {
            TotalSalesItem tsItem = new TotalSalesItem { State = context.State, Category = context.CategoryName, Product = context.ProductName };
            int y = DateTime.Today.Year - 1;
            for(int currentMonth = 1; currentMonth <= 12; currentMonth++) {
                DateTime dt = new DateTime(y, currentMonth, 1);
                context.UnitsSoldGenerator.Next();
                int uSold = context.UnitsSoldGenerator.UnitsSold;
                int uSoldTarget = context.UnitsSoldGenerator.UnitsSoldTarget;
                decimal rev = uSold * context.ListPrice;
                decimal revTarget = uSoldTarget * context.ListPrice;
                monthlySales.Add(new MonthlySalesItem {
                    State = context.State,
                    Product = context.ProductName,
                    Category = context.CategoryName,
                    CurrentDate = dt,
                    UnitsSold = uSold,
                    UnitsSoldTarget = uSoldTarget,
                    Revenue = rev,
                    RevenueTarget = revTarget
                });
                tsItem.RevenueYTD += rev;
                tsItem.RevenueYTDTarget += revTarget;
                tsItem.UnitsSoldYTD += uSold;
                tsItem.UnitsSoldYTDTarget += uSoldTarget;
                if(currentMonth >= 10 && currentMonth <= 12) {
                    tsItem.RevenueQTD += rev;
                    tsItem.RevenueQTDTarget += revTarget;
                }
                item.RevenueYTD += rev;
                item.RevenueYTDTarget += revTarget;
            }
            totalSales.Add(tsItem);
        }
        protected override void EndGenerate() {
            base.EndGenerate();
            item.ExpencesYTD = item.RevenueYTDTarget * 0.2m;
            item.ExpencesYTDTarget = item.RevenueYTDTarget * 0.1999m;
            item.ProfitYTD = item.RevenueYTD - item.ExpencesYTD;
            item.ProfitYTDTarget = item.RevenueYTDTarget - item.ExpencesYTDTarget;
            item.AvgOrderSizeYTD = item.RevenueYTD * 0.006m;
            item.AvgOrderSizeYTDTarget = item.RevenueYTDTarget * 0.0055m;
            item.NewCustomersYTD = (int)Math.Round(item.RevenueYTD * 0.0013m);
            item.NewCustomersYTDTarget = (int)Math.Round(item.RevenueYTDTarget * 0.00125m);
            item.MarketShare = 0.23f;
        }
    }
    public class UnitsSoldRandomGenerator {
        const int MinUnitsSold = 5;

        readonly Random rand;
        readonly int startUnitsSold;
        int? prevUnitsSold;
        int? prevPrevUnitsSold;
        int unitsSold;
        int unitsSoldTarget;
        bool isFirst = true;

        public int UnitsSold { get { return unitsSold; } }
        public int UnitsSoldTarget { get { return unitsSoldTarget; } }

        public UnitsSoldRandomGenerator(Random rand, int startUnitsSold) {
            this.rand = rand;
            this.startUnitsSold = Math.Max(startUnitsSold, MinUnitsSold);
        }
        public void Next() {
            if(isFirst) {
                unitsSold = startUnitsSold;
                isFirst = false;
            } else {
                unitsSold = unitsSold + (int)Math.Round(DataHelper.Random(rand, unitsSold * 0.5));
                unitsSold = Math.Max(unitsSold, MinUnitsSold);
            }
            int unitsSoldSum = unitsSold;
            int count = 1;
            if(prevUnitsSold.HasValue) {
                unitsSoldSum += prevUnitsSold.Value;
                count++;
            }
            if(prevPrevUnitsSold.HasValue) {
                unitsSoldSum += prevPrevUnitsSold.Value;
                count++;
            }
            unitsSoldTarget = (int)Math.Round((double)unitsSoldSum / count);
            unitsSoldTarget = unitsSoldTarget + (int)Math.Round(DataHelper.Random(rand, unitsSoldTarget));
            prevPrevUnitsSold = prevUnitsSold;
            prevUnitsSold = unitsSold;
        }
    }
    public class ProductClasses : List<ProductClass> {
        public new ProductClass this[int productID] {
            get {
                foreach(ProductClass productClass in this)
                    if(productClass.ContainsProduct(productID))
                        return productClass;
                throw new ArgumentException("procutID");
            }
        }
        public ProductClasses(ICollection products) {
            Add(new ProductClass(null, 100m, 0.5));
            Add(new ProductClass(100m, 500m, 0.4));
            Add(new ProductClass(500m, 1500m, 0.3));
            Add(new ProductClass(1500m, null, 0.2));
            foreach(DataRow product in products) {
                int productID = (int)product["ProductID"];
                decimal listPrice = (decimal)product["ListPrice"];
                foreach(ProductClass productClass in this)
                    if(productClass.AddProduct(productID, listPrice))
                        break;
            }
        }
    }
    public class RegionClasses : Dictionary<int, double> {
        public RegionClasses(ICollection regions) {
            int? numberEmployeesMin = null;
            foreach(DataRow region in regions) {
                short numberEmployees = (short)region["NumberEmployees"];
                numberEmployeesMin = numberEmployeesMin.HasValue ? Math.Min(numberEmployeesMin.Value, numberEmployees) : numberEmployees;
            }
            foreach(DataRow region in regions)
                Add((int)region["RegionID"], (short)region["NumberEmployees"] / (double)numberEmployeesMin.Value);
        }
    }
    public class ProductClass {
        readonly List<int> productIDs = new List<int>();
        readonly decimal? minPrice;
        readonly decimal? maxPrice;
        readonly double saleProbability;

        public double SaleProbability { get { return saleProbability; } }

        public ProductClass(decimal? minPrice, decimal? maxPrice, double saleProbability) {
            this.minPrice = minPrice;
            this.maxPrice = maxPrice;
            this.saleProbability = saleProbability;
        }
        public bool AddProduct(int productID, decimal price) {
            bool satisfyMinPrice = !minPrice.HasValue || price >= minPrice.Value;
            bool satisfyMaxPrice = !maxPrice.HasValue || price < maxPrice.Value;
            if(satisfyMinPrice && satisfyMaxPrice) {
                productIDs.Add(productID);
                return true;
            }
            return false;
        }
        public bool ContainsProduct(int productID) {
            return productIDs.Contains(productID);
        }
    }
    public static class DataHelper {
        public static double Random(Random random, double deviation, bool positive) {
            int rand = random.Next(positive ? 0 : -1000000, 1000000);
            return (double)rand / 1000000 * deviation;
        }
        public static double Random(Random random, double deviation) {
            return Random(random, deviation, false);
        }
    }
    public class TotalSalesItem {
        public string State { get; set; }
        public string Category { get; set; }
        public string Product { get; set; }
        public decimal RevenueYTD { get; set; }
        public decimal RevenueYTDTarget { get; set; }
        public decimal RevenueQTD { get; set; }
        public decimal RevenueQTDTarget { get; set; }
        public int UnitsSoldYTD { get; set; }
        public int UnitsSoldYTDTarget { get; set; }
    }
    public class MonthlySalesItem {
        public string State { get; set; }
        public string Product { get; set; }
        public string Category { get; set; }
        public DateTime CurrentDate { get; set; }
        public decimal Revenue { get; set; }
        public decimal RevenueTarget { get; set; }
        public int UnitsSold { get; set; }
        public int UnitsSoldTarget { get; set; }
    }
    public class KeyMetricsItem {
        public decimal RevenueYTD { get; set; }
        public decimal RevenueYTDTarget { get; set; }
        public decimal ExpencesYTD { get; set; }
        public decimal ExpencesYTDTarget { get; set; }
        public decimal ProfitYTD { get; set; }
        public decimal ProfitYTDTarget { get; set; }
        public decimal AvgOrderSizeYTD { get; set; }
        public decimal AvgOrderSizeYTDTarget { get; set; }
        public int NewCustomersYTD { get; set; }
        public int NewCustomersYTDTarget { get; set; }
        public float MarketShare { get; set; }
    }
}

File: DevExpress-Demo/VB/DockingDemo.Wpf/Helpers/DashboardDataHelper.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Data
Imports System.IO
Imports System.Reflection
Imports DevExpress.DemoData.Helpers
Imports DevExpress.Xpf.DemoBase.Helpers

Namespace DockingDemo.Helpers

    Public Module DataLoader

        Private Function LoadData(ByVal fileName As String) As DataSet
            Dim path As String = DevExpress.DemoData.Helpers.DataFilesHelper.FindFile(String.Format("{0}.xml", fileName), DevExpress.DemoData.Helpers.DataFilesHelper.DataPath)
            Dim ds As System.Data.DataSet = New System.Data.DataSet()
            ds.ReadXml(path, System.Data.XmlReadMode.ReadSchema)
            Return ds
        End Function

        Public Function LoadSales() As DataSet
            Return DockingDemo.Helpers.DataLoader.LoadData("DashboardSales")
        End Function
    End Module

    Public MustInherit Class SalesDataGenerator

        Public Class Context

            Private ReadOnly st As String

            Private ReadOnly prodName As String

            Private ReadOnly catName As String

            Private ReadOnly lPrice As Decimal

            Private ReadOnly uSoldGenerator As DockingDemo.Helpers.UnitsSoldRandomGenerator

            Public ReadOnly Property State As String
                Get
                    Return Me.st
                End Get
            End Property

            Public ReadOnly Property ProductName As String
                Get
                    Return Me.prodName
                End Get
            End Property

            Public ReadOnly Property CategoryName As String
                Get
                    Return Me.catName
                End Get
            End Property

            Public ReadOnly Property ListPrice As Decimal
                Get
                    Return Me.lPrice
                End Get
            End Property

            Public ReadOnly Property UnitsSoldGenerator As UnitsSoldRandomGenerator
                Get
                    Return Me.uSoldGenerator
                End Get
            End Property

            Public Sub New(ByVal st As String, ByVal prodName As String, ByVal catName As String, ByVal lPrice As Decimal, ByVal uSoldGenerator As DockingDemo.Helpers.UnitsSoldRandomGenerator)
                Me.st = st
                Me.prodName = prodName
                Me.catName = catName
                Me.lPrice = lPrice
                Me.uSoldGenerator = uSoldGenerator
            End Sub
        End Class

        Protected Shared Function GetState(ByVal region As System.Data.DataRow) As String
            Return CStr(region("Region"))
        End Function

        Protected Shared Function GetProductName(ByVal product As System.Data.DataRow) As String
            Return CStr(product("Name"))
        End Function

        Protected Shared Function GetListPrice(ByVal product As System.Data.DataRow) As Decimal
            Return CDec(product("ListPrice"))
        End Function

        Private ReadOnly ds As System.Data.DataSet

        Private ReadOnly categoriesTable As System.Data.DataTable

        Private ReadOnly productsTable As System.Data.DataTable

        Private ReadOnly regionsTable As System.Data.DataTable

        Private ReadOnly rand As System.Random = New System.Random(1)

        Private ReadOnly prodClassesField As DockingDemo.Helpers.ProductClasses

        Private ReadOnly regClassesField As DockingDemo.Helpers.RegionClasses

        Protected ReadOnly Property Regions As DataRowCollection
            Get
                Return Me.regionsTable.Rows
            End Get
        End Property

        Protected ReadOnly Property Products As DataRowCollection
            Get
                Return Me.productsTable.Rows
            End Get
        End Property

        Protected ReadOnly Property ProdClasses As ProductClasses
            Get
                Return Me.prodClassesField
            End Get
        End Property

        Protected ReadOnly Property RegClasses As RegionClasses
            Get
                Return Me.regClassesField
            End Get
        End Property

        Protected ReadOnly Property Random As Random
            Get
                Return Me.rand
            End Get
        End Property

        Protected Sub New(ByVal ds As System.Data.DataSet)
            Me.ds = ds
            Me.categoriesTable = ds.Tables("Categories")
            Me.productsTable = ds.Tables("Products")
            Me.regionsTable = ds.Tables("Regions")
            Me.prodClassesField = New DockingDemo.Helpers.ProductClasses(Me.productsTable.Rows)
            Me.regClassesField = New DockingDemo.Helpers.RegionClasses(Me.regionsTable.Rows)
        End Sub

        Protected Function GetRegionWeigtht(ByVal region As System.Data.DataRow) As Double
            Return Me.regClassesField(CInt(region("RegionID")))
        End Function

        Protected Function GetProductClass(ByVal product As System.Data.DataRow) As ProductClass
            Return Me.prodClassesField(CInt(product("ProductID")))
        End Function

        Protected Function GetCategoryName(ByVal product As System.Data.DataRow) As String
            Return CStr(Me.categoriesTable.[Select](String.Format("CategoryID = {0}", product("CategoryID")))(0)("CategoryName"))
        End Function

        Protected Function CreateUnitsSoldGenerator(ByVal regionWeight As Double, ByVal productClass As DockingDemo.Helpers.ProductClass) As UnitsSoldRandomGenerator
            Return New DockingDemo.Helpers.UnitsSoldRandomGenerator(Me.rand, CInt(System.Math.Ceiling(productClass.SaleProbability * regionWeight)))
        End Function

        Protected MustOverride Sub Generate(ByVal context As DockingDemo.Helpers.SalesDataGenerator.Context)

        Protected Overridable Sub EndGenerate()
        End Sub

        Public Sub Generate()
            For Each region As System.Data.DataRow In Me.Regions
                Dim state As String = DockingDemo.Helpers.SalesDataGenerator.GetState(region)
                Dim regionWeight As Double = Me.GetRegionWeigtht(region)
                For Each product As System.Data.DataRow In Me.Products
                    Dim unitsSoldgenerator As DockingDemo.Helpers.UnitsSoldRandomGenerator = Me.CreateUnitsSoldGenerator(regionWeight, Me.GetProductClass(product))
                    Me.Generate(New DockingDemo.Helpers.SalesDataGenerator.Context(state, DockingDemo.Helpers.SalesDataGenerator.GetProductName(product), Me.GetCategoryName(product), DockingDemo.Helpers.SalesDataGenerator.GetListPrice(product), unitsSoldgenerator))
                Next
            Next

            Me.EndGenerate()
        End Sub
    End Class

    Public Class SalesPerformanceDataGenerator
        Inherits DockingDemo.Helpers.SalesDataGenerator

        Private ReadOnly monthlySalesField As System.Collections.Generic.List(Of DockingDemo.Helpers.MonthlySalesItem) = New System.Collections.Generic.List(Of DockingDemo.Helpers.MonthlySalesItem)()

        Private ReadOnly totalSalesField As System.Collections.Generic.List(Of DockingDemo.Helpers.TotalSalesItem) = New System.Collections.Generic.List(Of DockingDemo.Helpers.TotalSalesItem)()

        Private ReadOnly item As DockingDemo.Helpers.KeyMetricsItem = New DockingDemo.Helpers.KeyMetricsItem()

        Public ReadOnly Property MonthlySales As IEnumerable(Of DockingDemo.Helpers.MonthlySalesItem)
            Get
                Return Me.monthlySalesField
            End Get
        End Property

        Public ReadOnly Property TotalSales As IEnumerable(Of DockingDemo.Helpers.TotalSalesItem)
            Get
                Return Me.totalSalesField
            End Get
        End Property

        Public ReadOnly Property KeyMetrics As IEnumerable(Of DockingDemo.Helpers.KeyMetricsItem)
            Get
                Return New DockingDemo.Helpers.KeyMetricsItem() {Me.item}
            End Get
        End Property

        Public Sub New(ByVal dataSet As System.Data.DataSet)
            MyBase.New(dataSet)
        End Sub

        Protected Overrides Sub Generate(ByVal context As DockingDemo.Helpers.SalesDataGenerator.Context)
            Dim tsItem As DockingDemo.Helpers.TotalSalesItem = New DockingDemo.Helpers.TotalSalesItem With {.State = context.State, .Category = context.CategoryName, .Product = context.ProductName}
            Dim y As Integer = System.DateTime.Today.Year - 1
            For currentMonth As Integer = 1 To 12
                Dim dt As System.DateTime = New System.DateTime(y, currentMonth, 1)
                context.UnitsSoldGenerator.[Next]()
                Dim uSold As Integer = context.UnitsSoldGenerator.UnitsSold
                Dim uSoldTarget As Integer = context.UnitsSoldGenerator.UnitsSoldTarget
                Dim rev As Decimal = uSold * context.ListPrice
                Dim revTarget As Decimal = uSoldTarget * context.ListPrice
                Me.monthlySalesField.Add(New DockingDemo.Helpers.MonthlySalesItem With {.State = context.State, .Product = context.ProductName, .Category = context.CategoryName, .CurrentDate = dt, .UnitsSold = uSold, .UnitsSoldTarget = uSoldTarget, .Revenue = rev, .RevenueTarget = revTarget})
                tsItem.RevenueYTD += rev
                tsItem.RevenueYTDTarget += revTarget
                tsItem.UnitsSoldYTD += uSold
                tsItem.UnitsSoldYTDTarget += uSoldTarget
                If currentMonth >= 10 AndAlso currentMonth <= 12 Then
                    tsItem.RevenueQTD += rev
                    tsItem.RevenueQTDTarget += revTarget
                End If

                Me.item.RevenueYTD += rev
                Me.item.RevenueYTDTarget += revTarget
            Next

            Me.totalSalesField.Add(tsItem)
        End Sub

        Protected Overrides Sub EndGenerate()
            MyBase.EndGenerate()
            Me.item.ExpencesYTD = Me.item.RevenueYTDTarget * 0.2D
            Me.item.ExpencesYTDTarget = Me.item.RevenueYTDTarget * 0.1999D
            Me.item.ProfitYTD = Me.item.RevenueYTD - Me.item.ExpencesYTD
            Me.item.ProfitYTDTarget = Me.item.RevenueYTDTarget - Me.item.ExpencesYTDTarget
            Me.item.AvgOrderSizeYTD = Me.item.RevenueYTD * 0.006D
            Me.item.AvgOrderSizeYTDTarget = Me.item.RevenueYTDTarget * 0.0055D
            Me.item.NewCustomersYTD = CInt(System.Math.Round(Me.item.RevenueYTD * 0.0013D))
            Me.item.NewCustomersYTDTarget = CInt(System.Math.Round(Me.item.RevenueYTDTarget * 0.00125D))
            Me.item.MarketShare = 0.23F
        End Sub
    End Class

    Public Class UnitsSoldRandomGenerator

        Const MinUnitsSold As Integer = 5

        Private ReadOnly rand As System.Random

        Private ReadOnly startUnitsSold As Integer

        Private prevUnitsSold As Integer?

        Private prevPrevUnitsSold As Integer?

        Private unitsSoldField As Integer

        Private unitsSoldTargetField As Integer

        Private isFirst As Boolean = True

        Public ReadOnly Property UnitsSold As Integer
            Get
                Return Me.unitsSoldField
            End Get
        End Property

        Public ReadOnly Property UnitsSoldTarget As Integer
            Get
                Return Me.unitsSoldTargetField
            End Get
        End Property

        Public Sub New(ByVal rand As System.Random, ByVal startUnitsSold As Integer)
            Me.rand = rand
            Me.startUnitsSold = System.Math.Max(startUnitsSold, DockingDemo.Helpers.UnitsSoldRandomGenerator.MinUnitsSold)
        End Sub

        Public Sub [Next]()
            If Me.isFirst Then
                Me.unitsSoldField = Me.startUnitsSold
                Me.isFirst = False
            Else
                Me.unitsSoldField = Me.unitsSoldField + CInt(System.Math.Round(DockingDemo.Helpers.DataHelper.Random(Me.rand, Me.unitsSoldField * 0.5)))
                Me.unitsSoldField = System.Math.Max(Me.unitsSoldField, DockingDemo.Helpers.UnitsSoldRandomGenerator.MinUnitsSold)
            End If

            Dim unitsSoldSum As Integer = Me.unitsSoldField
            Dim count As Integer = 1
            If Me.prevUnitsSold.HasValue Then
                unitsSoldSum += Me.prevUnitsSold.Value
                count += 1
            End If

            If Me.prevPrevUnitsSold.HasValue Then
                unitsSoldSum += Me.prevPrevUnitsSold.Value
                count += 1
            End If

            Me.unitsSoldTargetField = CInt(System.Math.Round(CDbl(unitsSoldSum) / count))
            Me.unitsSoldTargetField = Me.unitsSoldTargetField + CInt(System.Math.Round(DockingDemo.Helpers.DataHelper.Random(Me.rand, Me.unitsSoldTargetField)))
            Me.prevPrevUnitsSold = Me.prevUnitsSold
            Me.prevUnitsSold = Me.unitsSoldField
        End Sub
    End Class

    Public Class ProductClasses
        Inherits System.Collections.Generic.List(Of DockingDemo.Helpers.ProductClass)

        Default Public Overloads ReadOnly Property Item(ByVal productID As Integer) As ProductClass
            Get
                For Each productClass As DockingDemo.Helpers.ProductClass In Me
                    If productClass.ContainsProduct(productID) Then Return productClass
                Next

                Throw New System.ArgumentException("procutID")
            End Get
        End Property

        Public Sub New(ByVal products As System.Collections.ICollection)
            Me.Add(New DockingDemo.Helpers.ProductClass(Nothing, 100D, 0.5))
            Me.Add(New DockingDemo.Helpers.ProductClass(100D, 500D, 0.4))
            Me.Add(New DockingDemo.Helpers.ProductClass(500D, 1500D, 0.3))
            Me.Add(New DockingDemo.Helpers.ProductClass(1500D, Nothing, 0.2))
            For Each product As System.Data.DataRow In products
                Dim productID As Integer = CInt(product("ProductID"))
                Dim listPrice As Decimal = CDec(product("ListPrice"))
                For Each productClass As DockingDemo.Helpers.ProductClass In Me
                    If productClass.AddProduct(productID, listPrice) Then Exit For
                Next
            Next
        End Sub
    End Class

    Public Class RegionClasses
        Inherits System.Collections.Generic.Dictionary(Of Integer, Double)

        Public Sub New(ByVal regions As System.Collections.ICollection)
            Dim numberEmployeesMin As Integer? = Nothing
            For Each region As System.Data.DataRow In regions
                Dim numberEmployees As Short = CShort(region("NumberEmployees"))
                numberEmployeesMin = If(numberEmployeesMin.HasValue, System.Math.Min(numberEmployeesMin.Value, numberEmployees), numberEmployees)
            Next

            For Each region As System.Data.DataRow In regions
                Me.Add(CInt(region("RegionID")), CShort(region("NumberEmployees")) / CDbl(numberEmployeesMin.Value))
            Next
        End Sub
    End Class

    Public Class ProductClass

        Private ReadOnly productIDs As System.Collections.Generic.List(Of Integer) = New System.Collections.Generic.List(Of Integer)()

        Private ReadOnly minPrice As Decimal?

        Private ReadOnly maxPrice As Decimal?

        Private ReadOnly saleProbabilityField As Double

        Public ReadOnly Property SaleProbability As Double
            Get
                Return Me.saleProbabilityField
            End Get
        End Property

        Public Sub New(ByVal minPrice As Decimal?, ByVal maxPrice As Decimal?, ByVal saleProbability As Double)
            Me.minPrice = minPrice
            Me.maxPrice = maxPrice
            Me.saleProbabilityField = saleProbability
        End Sub

        Public Function AddProduct(ByVal productID As Integer, ByVal price As Decimal) As Boolean
            Dim satisfyMinPrice As Boolean = Not Me.minPrice.HasValue OrElse price >= Me.minPrice.Value
            Dim satisfyMaxPrice As Boolean = Not Me.maxPrice.HasValue OrElse price < Me.maxPrice.Value
            If satisfyMinPrice AndAlso satisfyMaxPrice Then
                Me.productIDs.Add(productID)
                Return True
            End If

            Return False
        End Function

        Public Function ContainsProduct(ByVal productID As Integer) As Boolean
            Return Me.productIDs.Contains(productID)
        End Function
    End Class

    Public Module DataHelper

        Public Function Random(ByVal pRandom As System.Random, ByVal deviation As Double, ByVal positive As Boolean) As Double
            Dim rand As Integer = pRandom.[Next](If(positive, 0, -1000000), 1000000)
            Return CDbl(rand) / 1000000 * deviation
        End Function

        Public Function Random(ByVal pRandom As System.Random, ByVal deviation As Double) As Double
            Return DockingDemo.Helpers.DataHelper.Random(pRandom, deviation, False)
        End Function
    End Module

    Public Class TotalSalesItem

        Public Property State As String

        Public Property Category As String

        Public Property Product As String

        Public Property RevenueYTD As Decimal

        Public Property RevenueYTDTarget As Decimal

        Public Property RevenueQTD As Decimal

        Public Property RevenueQTDTarget As Decimal

        Public Property UnitsSoldYTD As Integer

        Public Property UnitsSoldYTDTarget As Integer
    End Class

    Public Class MonthlySalesItem

        Public Property State As String

        Public Property Product As String

        Public Property Category As String

        Public Property CurrentDate As DateTime

        Public Property Revenue As Decimal

        Public Property RevenueTarget As Decimal

        Public Property UnitsSold As Integer

        Public Property UnitsSoldTarget As Integer
    End Class

    Public Class KeyMetricsItem

        Public Property RevenueYTD As Decimal

        Public Property RevenueYTDTarget As Decimal

        Public Property ExpencesYTD As Decimal

        Public Property ExpencesYTDTarget As Decimal

        Public Property ProfitYTD As Decimal

        Public Property ProfitYTDTarget As Decimal

        Public Property AvgOrderSizeYTD As Decimal

        Public Property AvgOrderSizeYTDTarget As Decimal

        Public Property NewCustomersYTD As Integer

        Public Property NewCustomersYTDTarget As Integer

        Public Property MarketShare As Single
    End Class
End Namespace
